### PR TITLE
docs: various updates to the Dockerfile reference

### PR DESCRIPTION
**Pull request by thaJeztah**
Still quite some things to do, but picking up some changes;

### docs/reference/builder: touch-up code-hints and some minor changes

- use "console" for code-hints, to make process output distinguishable from the commands that are executed
- use a consistent prompt for powershell examples
- minor changes in wording around "build context" to reduce confusion  with `docker context`

### docs/reference/builder: remove outdated example Dockerfiles

These examples were really outdated, so linking to other sections in the documentation instead.

### docs/reference/builder: update example output, and some rephrasing

- update some examples to show the BuildKit output
- remove some wording about "images" being used for the build cache
- add a link to the `--cache-from` section
- added a link to "scanning your image with `docker scan`"
- updated link to "push your image"

### docs/reference/builder: update "syntax" section

- rename "experimental" to "labs"
- rephrase recommendation for picking a version
- clarify that the "labs" channel provides a superset of the "stable" channel.
- remove "External implementation features" section, because it overlapped  with the "syntax" section.
- removed `:latest` from the "stable" channel (generally not recommended)

### CODE REVIEW

File: docs/reference/builder.md [hunk 1]
```diff
@@ -66,32 +67,37 @@ Traditionally, the `Dockerfile` is called `Dockerfile` and located in the root
...
 
-Sending build context to Docker daemon 2.048 kB
-Error response from daemon: Unknown instruction: RUNCMD
+[+] Building 0.3s (2/2) FINISHED
+ => [internal] load build definition from Dockerfile                       0.1s
+ => => transferring dockerfile: 60B                                        0.0s
+ => [internal] load .dockerignore                                          0.1s
+ => => transferring context: 2B                                            0.0s
+error: failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to create LLB definition:
+dockerfile parse error line 2: unknown instruction: RUNCMD
```
Comment (tonistiigi) [inline docs/reference/builder.md:100]: you can use the error after `buildx install`

File: docs/reference/builder.md [hunk 2]
```diff
@@ -104,38 +110,35 @@ Note that each instruction is run independently, and causes a new image
...
+ => => naming to docker.io/library/mysocatimage                            0.0s
 ```
 
-Build cache is only used from images that have a local parent chain. This means
-that these images were created by previous builds or the whole chain of images
-was loaded with `docker load`. If you wish to use build cache of a specific
-image you can specify it with `--cache-from` option. Images specified with
-`--cache-from` do not need to have a parent chain and may be pulled from other
-registries.
+By default, the build cache is based results from previous builds on the machine
```
Comment (tonistiigi) [inline docs/reference/builder.md:134]: "based on"

File: docs/reference/builder.md [hunk 3]
```diff
@@ -315,62 +318,80 @@ The following parser directives are supported:
...
+Custom Dockerfile implementations allows you to:
 
-  - Automatically get bugfixes without updating the daemon
+  - Automatically get bugfixes without updating the Docker daemon
   - Make sure all users are using the same implementation to build your Dockerfile
-  - Use the latest features without updating the daemon
-  - Try out new experimental or third-party features
+  - Use the latest features without updating the Docker daemon
+  - Try out new features or third-party features before they are integrated in the Docker daemon
+  - Use [alternative syntaxes, or create your own](https://github.com/moby/buildkit#exploring-llb)
```
Comment (tonistiigi) [inline docs/reference/builder.md:349]: syntaxes -> build definitions

File: docs/reference/builder.md [hunk 4]
```diff
@@ -315,62 +318,80 @@ The following parser directives are supported:
...
+syntax when performing a build, making sure you are using the most current version.
 
-  - `docker/dockerfile:1.0.0` - only allow immutable version `1.0.0`
-  - `docker/dockerfile:1.0` - allow versions `1.0.*`
-  - `docker/dockerfile:1` - allow versions `1.*.*`
-  - `docker/dockerfile:latest` - latest release on stable channel
+While it is possible to use more specific versions, such as `1.2` or `1.2.1`,
+doing so is generally not recommended as their versions should be updated to
+continue receiving updates. For example, `docker/dockerfile:1.2` stops receiving
+updates once version `1.3.0` is released.
```
Comment (tonistiigi) [inline docs/reference/builder.md:371]: If a specific version is used the Dockerfile needs to upgrade the version manually to continue receiving bugfixes and new features. The old versions remain compatible with the new versions of the builder.

File: docs/reference/builder.md [hunk 5]
```diff
@@ -315,62 +318,80 @@ The following parser directives are supported:
...
+of the features in the stable channel; note that `stable` features in the labs
+channel images follow [semantic versioning](https://semver.org), but "labs"
+features do not, and newer releases may not be backwards compatible, so it
 is recommended to use an immutable full version variant.
 
-For master builds and nightly feature releases refer to the description in
-[the source repository](https://github.com/moby/buildkit/blob/master/README.md).
+For documentation on "labs" features, master builds, and nightly feature releases,
+refer to the description in [the BuildKit source repository on GitHub](https://github.com/moby/buildkit/blob/master/README.md).
+For a full list of available images, visit the [image repository on Docker Hub](https://hub.docker.com/r/docker/dockerfile).
```
Comment (tonistiigi) [inline docs/reference/builder.md:394]: can also link to https://hub.docker.com/r/docker/dockerfile-upstream for development builds

File: docs/reference/builder.md (hunk 1 above)
Comment (thaJeztah) [reply docs/reference/builder.md:100]: You mean the improved errors? I tried with buildx 0.5.1, but that still showed this error; let me try installing buildx master

Comment (thaJeztah) [reply docs/reference/builder.md:100]: Same with buildx master and `docker/dockerfile-upstream:master`

```bash
docker buildx version
github.com/docker/buildx v0.5.1-77-g8d64b64 8d64b6484f9681a71233d439896b949064f3921c

docker buildx install

docker build -t test/myapp -<<EOF
# syntax=docker/dockerfile-upstream:master
FROM alpine:3.2
RUNCMD
EOF
```

Which produces:

```
error: failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to solve with frontend gateway.v0: rpc error: code = Unknown desc = failed to create LLB definition: dockerfile parse error line 3: unknown instruction: RUNCMD
```

File: docs/reference/builder.md (hunk 5 above)
Comment (thaJeztah) [reply docs/reference/builder.md]: added

File: docs/reference/builder.md (hunk 4 above)
Comment (thaJeztah) [reply docs/reference/builder.md]: Added a variation of the above; check if it looks ok

**Review by chris-crone — APPROVED** [review]
LGTM

**Review by tonistiigi — APPROVED** [review]

**Pull request merged**